flang/lib/Lower/OpenMP.cpp

```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
     auto masterOp =
         firOpBuilder.create<mlir::omp::MasterOp>(currentLocation, argTy);
     createBodyOfOp<omp::MasterOp>(masterOp, firOpBuilder, currentLocation);
   }
 }
 
+static void genOMP(Fortran::lower::AbstractConverter &converter,
+                   Fortran::lower::pft::Evaluation &eval,
+                   const Fortran::parser::OpenMPLoopConstruct &loopConstruct) {
+
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  auto currentLocation = converter.getCurrentLocation();
+  SmallVector<Value, 4> lowerBound, upperBound, step, privateClauseOperands,
+      firstPrivateClauseOperands, lastPrivateClauseOperands, linearVars,
+      linearStepVars;
+  mlir::Value scheduleChunkClauseOperand;
+  Attribute scheduleClauseOperand, collapseClauseOperand, noWaitClauseOperand,
+      orderedClauseOperand, orderClauseOperand;
+  const auto &wsLoopOpClauseList = std::get<Fortran::parser::OmpClauseList>(
+      std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t);
+  if (llvm::omp::OMPD_parallel_do ==
+      std::get<Fortran::parser::OmpLoopDirective>(
+          std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t)
+          .v)
+    createParallelOp<Fortran::parser::OmpBeginLoopDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t));
+  else {
+    for (const auto &clause : wsLoopOpClauseList.v) {
+      if (const auto &privateClause =
+              std::get_if<Fortran::parser::OmpClause::Private>(&clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = privateClause->v;
+        genObjectList(ompObjectList, converter, privateClauseOperands);
+      } else if (const auto &firstPrivateClause =
+                     std::get_if<Fortran::parser::OmpClause::Firstprivate>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList =
+            firstPrivateClause->v;
+        genObjectList(ompObjectList, converter, firstPrivateClauseOperands);
+      }
+    }
+  }
+  for (const auto &clause : wsLoopOpClauseList.v) {
+    if (const auto &lastPrivateClause =
+            std::get_if<Fortran::parser::OmpClause::Lastprivate>(&clause.u)) {
+      const Fortran::parser::OmpObjectList &ompObjectList =
+          lastPrivateClause->v;
+      genObjectList(ompObjectList, converter, lastPrivateClauseOperands);
+    }
+  }
+  auto &doStmtEval = eval.getFirstNestedEvaluation();
+  auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
+  const auto &loopControl =
+      std::get<std::optional<Fortran::parser::LoopControl>>(doStmt->t);
+  if (const auto *bounds =
+          std::get_if<Fortran::parser::LoopControl::Bounds>(&loopControl->u)) {
+    Fortran::lower::StatementContext stmtCtx;
+    lowerBound.push_back(fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(bounds->lower), stmtCtx)));
+    upperBound.push_back(fir::getBase(converter.genExprValue(
```

Comment:
At the moment the lowering of the wsloop is not inclusive of the upperbound. We have to add an attribute (inclusive?) to the operation to include the upperbound.

Reply:
The attribute is now in upstream llvm-project (see commit below). You have to either use that attribute or create an expression to increment the upperbound.
https://github.com/llvm/llvm-project/commit/268ff38a716157c362b8d463e2e5655f25972e42

Reply:
@schweitzpgi are there plans for a merge from upstream soon? If not, is it possible to cherry-pick the following commit?https://github.com/llvm/llvm-project/commit/268ff38a716157c362b8d463e2e5655f25972e42

Reply:
The next rebase should be happening later this week.